build/builder.go

```diff
@@ -350,8 +363,28 @@ func (o *Options) IndexState() (IndexState, string) {
 		return IndexStateOption, fn
 	}
 
-	if !reflect.DeepEqual(repo.Branches, o.RepositoryDescription.Branches) {
-		return IndexStateContent, fn
+	if o.IsDelta {
+		// TODO: Get rid of this guard once the delta shard behavior is the default
+		ignoreVersionOption := cmpopts.IgnoreFields(zoekt.RepositoryBranch{}, "Version")
+		sortBranchesOption := cmpopts.SortSlices(func(a, b zoekt.RepositoryBranch) bool {
+			if a.Name < b.Name {
+				return true
+			}
+
+			return a.Version < b.Version
+		})
+
+		if !cmp.Equal(repo.Branches, o.RepositoryDescription.Branches, ignoreVersionOption, sortBranchesOption) {
+			return IndexStateBranchSet, fn
+		}
+
+		if !cmp.Equal(repo.Branches, o.RepositoryDescription.Branches, sortBranchesOption) {
+			return IndexStateBranchVersion, fn
+		}
+	} else {
+		if !reflect.DeepEqual(repo.Branches, o.RepositoryDescription.Branches) {
+			return IndexStateContent, fn
+		}
 	}
 
 	// We can mutate repo since it lives in the scope of this function call.
```

Comment:
cc @stefanhengl I decided to not add the more granular branch indexstate information to MergeMutable since the only use of this is for compound shards. Since we're punting any work on that for now - no delta shard should end up in a compound shard (and the coarse `branches` check in merge mutable ensures that the function would error anyway), it doesn't seem necessary here